keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,393 @@
+	NetworkStats network = 4;
+	// Note the specific omission of VolumeStats and EphemeralStorage
+	// Each of these fields will be calculated Kubelet-level
+	// ProcessStats pertaining to processes.
+	ProcessStats process = 5;
+}
+
+// NetworkStats contains data about network resources.
+message NetworkStats {
+	// The time at which these stats were updated.
+	int64 timestamp = 1;
+
+	// Stats for the default interface, if found
+	InterfaceStats default_interface = 2;
+
+	repeated InterfaceStats interfaces = 3;
+}
+
+// InterfaceStats contains resource value data about interface.
+type InterfaceStats struct {
+	// The name of the interface
+	string name = 1;
+	// Cumulative count of bytes received.
+	Uint64Value rx_bytes = 2;
+	// Cumulative count of receive errors encountered.
+	Uint64Value rx_errors = 2;
+	// Cumulative count of bytes transmitted.
+	Uint64Value tx_bytes = 2;
+	// Cumulative count of transmit errors encountered.
+	Uint64Value tx_errors = 2;
+}
+
+// ProcessStats are stats pertaining to processes.
+message ProcessStats {
+	// Number of processes
+	Uint64Value process_count = 1;
+}
+```
+
+#### Kubelet
+
+Once all required CRI changes are completed, Kubelet can update it's CRI stats provider to stop fetching metrics from cAdvisor and instead obtain the metrics from the CRI for container and pods.
+
+* TODO: How to handle migration to newly created "full" CRI stats provider?
+    * Option #1: Modify existing CRI stats provider by removing all usage of cAdvisor. May lead to backwards compatibility issue due to possible differences between CRI and cAdvisor metric collection
+    * Option #2: Create a new stats provider backed by a feature gate that provides backward compatibility and allows transition to fully CRI backed stats provider 
+
+Additional work may be required to evaluate other kubelet components (e.g. eviction, preemption, etc) that may be relying on container or pod level metrics. Ideally all components will rely on summary API thereby alleviating need for cAdvisor for container and pod level stats. This is also a requirement to be able to disable cAdvisor container metrics collection.
+
+#### cAdvisor
+
+Once CRI and Kubelet stats provider level changes are in place, we can evaluate disabling cAdvisor from collecting container and pod level stats. We may need to introduce new ability in cAdvisor to disable on-demand collection for certain cgroup hierarchies to ensure we can continue using cAdvisor for only node/machine level stats. 
+
+### User Stories [optional]
+
+#### Story 1
+As a cluster admin, I would like my node to be as performant as possible, and not waste resources by duplicating stats collection by CRI and cAdvisor.
+
+#### Story 2
+As a node maintainer, I would like to clarify the sources of truth for each stat the Kubelet reports to the Summary API, so that I can better find which component is to investigate for any issues.
+
+
+### Risks and Mitigations
```

Comment:
Can we capture performance implications and any testing that will be done as part of this?

Reply:
added

Reply:
I put the responsibility of testing on CRI implementations, as we're going to be the new owners, and thus the ones responsible for not regressing